Allow support for multiple sites concurrently w/ `commerce-sdk-react`

bendvc opened the pull request:
![image](https://user-images.githubusercontent.com/8902581/213764277-0b4a056c-6bb4-4a7e-9e00-59ac1b74043a.png)

> Prefix your cookies with milk

# Description

To be in alignment with `plugin_slas` this PR ensures that the storage keys we use during the authentication process are prefixed with the id of the ecomm back-end we are connecting to. This means that you can jump from one site to another without loosing your guest or registered session, but more importantly the cookie keys will be the same format as what `plugin_slas` expects, so there will be no issues navigating from PWA to SFRA experiences.

# Changes

- Make BaseStorage an abstract class instead of an interface to add prefixing functionality.
- Remove clearing of storage as it's not in use anymore.
- Add new storage class for server side storage instead of using Maps directly (solves typing issue).

# How to Test-Drive This PR

- In the `test-commerce-sdk-react` project execute the command `npm start`.
- Goto the `slas login helpers` page. 
- Open your inspector and goto the application --> cookies tab and look at the cookies being created.
- You should see the `cc-nx-g` cookie prefixed with the currently selected site. 
- Try changing the site, you should see an additional cookie created with the newly selected site's id prefix to it.
- Now try logging in as a registered user, the guest cookie (`<prefix>_cc-nx-g`) will be replaces with the registered cooke (`<prefix>_cc-nx`)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [x] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

bendvc committed:
Initial PR commit.

bendvc committed:
First pass at implementation

bendvc committed:
Merge branch 'develop' into namespace-auth-storage

bendvc committed:
Remove clearStorage function as it's not in use

bendvc committed:
Make `getPrefixedKey` protected and `map` private

bendvc committed:
Add tests for new storage class

bendvc committed:
Merge branch 'namespace-auth-storage' of https://github.com/SalesforceCommerceCloud/pwa-kit into namespace-auth-storage

bendvc committed:
Make test easier to make additions too

bendvc committed:
Lint!

bendvc commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -136,39 +135,35 @@ class Auth {
-                expires: this.REFRESH_TOKEN_EXPIRATION_DAYS
-            })
-        }
+        // NOTE: Not too sure how I feel about this, seems a little confusing. Should
```
Thoughts on this would be appreciated. It isn't horrible, but I think it could be cleaner.

kevinxh commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -41,33 +62,64 @@ export class CookieStorage implements BaseStorage {
+    constructor(options?: BaseStorageOptions) {
+        super(options)
+
+        if (typeof window !== 'undefined') {
+            throw new Error('ServerStorage is not avaliable on the current environment.')
+        }
```
The implementation of `ServerStorage` runs totally fine on a browser environment, I don't think we need this error message? 

I wonder if `MemoryStorage` is a better name for this...

bendvc replied:
Good idea. I was thinking the same.

kevinxh commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -6,10 +6,26 @@
+    keyPrefixSeperator?: string
+}
+
+export abstract class BaseStorage {
```
An abstract class with implementation, fancy typescript stuff huh.

bendvc replied:
Ha.. had to dust of an old bank in my memory to remember that stuff from java

kevinxh commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -136,39 +135,35 @@ class Auth {
+        // Another idea is to implement a getPrefix function on the storage class. But that
+        // would look kinda gross too.
+        const storageOptions = {keyPrefix: config.siteId}
+        this.stores = onClient
+            ? {
+                  cookie: new CookieStorage(storageOptions),
+                  local: new LocalStorage(storageOptions)
+              }
+            : {
+                  cookie: new ServerStorage(storageOptions),
+                  local: new ServerStorage(storageOptions)
+              }
```
I wanna call out and discuss the way we initialize the storages...

Prior to this change, we initialize the storage outside of the `Auth` class, globally, and the reason for that is we want to the token to persist on server side across lambda invocations. As we all know, two serial SLAS calls for every single visit on the server side is expensive. (I also noticed a bug with the previous implementation, that the storage won't properly refresh when config value change, except `siteId`)

After this change, the storages are bound to the instance of the Auth class. Right now, we re-initialize Auth class in the provider when config changes.

```jsx
// We do this in provider
const auth = useMemo(() => {
        return new Auth({
            clientId,
            organizationId,
            shortCode,
            siteId,
            proxy,
            redirectURI,
            fetchOptions
        })
    }, [clientId, organizationId, shortCode, siteId, proxy, redirectURI, fetchOptions])
```

When a new Auth instance is created, the storages are wiped. I think this is expected behavior and fixed the bug i mentioned above. But we lose the ability to cache token on lambda, what can we do to keep caching token on lambda? 🤔

bendvc replied:
So to maintain the desired behaviour of retaining the token storage on the server side between lambda runs, I've added the ability for the MemoryStorage class to use a global map if required (which it is for our case), I called this option `sharedContext` so that one Auth class instance will use the same data storage as another Auth class instance.

kevinxh commented:
We probably want some sort of test that make sure the multi-site feature is working.

Particularly two test cases
1. I switch site, and things still works
2. I switch site, and switch back, i can use the old token

Any idea where these kind of tests belong to?

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -6,10 +6,26 @@
+    }
+    abstract set(key: string, value: string, options?: unknown): void
+    abstract get(key: string): string
+    abstract delete(key: string): void
```
We could potentially return `boolean` ("was deleted?") to fully conform with the [`Map` interface](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/delete#return_value). Not important if we don't, though.

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -41,33 +62,64 @@ export class CookieStorage implements BaseStorage {
 }
+
+export class ServerStorage extends BaseStorage {
+    private map: Map<string, string>
```
You can use fancy modern syntax!
```suggestion
    private map = Map<string, string>()
```
(Plus delete line 111)

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.test.ts:
```diff
@@ -0,0 +1,58 @@
+
+            if (!requiresWindow) {
+                global.window = window
+            }
```
```suggestion
            if (requiresWindow) {
                global.window = window
            }
```

bendvc replied:
This code has been removed as the code was iterated on.

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -6,10 +6,26 @@
+    protected getPrefixedKey(key: string): string {
+        const {keyPrefix, keyPrefixSeperator = '_'} = this?.options || {}
+        return `${keyPrefix ? keyPrefix + keyPrefixSeperator : ''}${key}`
+    }
```
```suggestion
    protected options: Required<BaseStorageOptions>

    constructor(options: BaseStorageOptions = {keyPrefixSeparator: '_'}) {
        this.options = {
          keyPrefixSeparator: options.keyPrefix ? options.keyPrefixSeparator ?? '_' : ''
          keyPrefix: options.keyPrefix ?? ''
        }
    }

    protected getPrefixedKey(key: string): string {
        return `${this.keyPrefix}${this.keyPrefixSeparator}${key}`
    }
```
1. Makes life simpler to just do it once, rather than every call. 😉 
2. Typo - it's sep**a**rator
3. Default param isn't necessary, since we need the fallback in the code anyway, but it could be helpful for autocomplete / generated docs.

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -6,10 +6,26 @@
-    set(key: string, value: string, options?: unknown): void
-    get(key: string): string
-    delete(key: string): void
+export type StorageTypes = 'cookie' | 'local'
```
```suggestion
export type StorageTypes = 'cookie' | 'local' | 'memory'
```
(Assuming we use the suggested `MemoryStorage` name.)

Even if we don't use `memory` ourselves, would adding this make sense?

bendvc replied:
Nice catch. 🙇

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -6,10 +6,26 @@
-    set(key: string, value: string, options?: unknown): void
-    get(key: string): string
-    delete(key: string): void
+export type StorageTypes = 'cookie' | 'local'
```
```suggestion
export type StorageType = 'cookie' | 'local'
```
I feel like this kind of type is conventionally singular, but I could be wrong... 🤔

bendvc replied:
I agree.

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -6,10 +6,26 @@
-    set(key: string, value: string, options?: unknown): void
-    get(key: string): string
-    delete(key: string): void
+export type StorageTypes = 'cookie' | 'local'
```
🤔 We could add `static type: StorageType` to each implementation. That could potentially enable some fancy type magic. Not sure we have a use case for fancy type magic here, but maybe someone downstream will...

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.test.ts:
```diff
@@ -0,0 +1,58 @@
+            const storage = new ServerStorage(storageOptions)
+            validate(storage)
+
+            if (!requiresWindow) {
+                global.window = window
+            }
```
Don't tests that fail throw? This block would never get reached when tests fail. Seems like `window` management should be done in `before`/`after` blocks.

bendvc replied:
This is dead code now.

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.test.ts:
```diff
@@ -0,0 +1,58 @@
+            storage.set(key, value)
+            expect(storage.get(key)).toBe(value)
+            // @ts-ignore
+            expect(Array.from(storage.map.entries())[0][0]).toBe(`prefix$${key}`)
```
```suggestion
            expect([...storage.map.keys()]).toEqual([`prefix$${key}`])
```
Just a touch shorter. Also, subtly double checks that there's exactly one key!

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.test.ts:
```diff
@@ -0,0 +1,58 @@
+
+const testCases = [
+    {
+        description: 'ServerStorage works without options',
```
```suggestion
        description: 'ServerStorage works with empty options',
```
"without" to me implies `undefined`, not empty object.

bendvc replied:
I went the other way here and set the `storageOptions` to `undefined` as that is actually what I meant to test.

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.test.ts:
```diff
@@ -0,0 +1,58 @@
+ */
+import {BaseStorage, ServerStorage} from './storage'
+
+const {window} = global
```
Does jest inject this somehow? In basic node env, `window` will be `undefined`.

bendvc replied:
In the `pwa-kit-dev` module, you'll see that the environment that we are using for jest is `jsdom` in the `/src/config/jest/jest.config.js` file. 
Whether or not this is the correct environment for be defaulted to for this package is a question that we can think about, but maybe we can take that outside of the scope of this PR.

wjhsf commented on packages/commerce-sdk-react/src/auth/index.ts:
```diff
@@ -41,15 +41,13 @@ type AuthDataKeys =
 type AuthDataMap = Record<
```
Not relevant to this PR, but I was curious if there was a way to use `TokenResponse` so that we don't need to manually list the keys. Turns out there is!

```ts
// Helpers
type RemoveNeverValues<T> = {
    // Key remapping, fancy stuff! https://devblogs.microsoft.com/typescript/announcing-typescript-4-1/#key-remapping-in-mapped-types
    [K in keyof T as T[K] extends never ? never: K]: T[K]
}
type StringIndexNever<T> = {
    [K in keyof T]: string extends K ? never : T[K]
}
type KnownKeys<T> = keyof RemoveNeverValues<StringIndexNever<T>>

// Implementation
type TokenResponseKeys = KnownKeys<ShopperLoginTypes.TokenResponse>
type AuthDataKeys =
    | Exclude<TokenResponseKeys, 'refresh_token'>
    | 'idp_access_token'
    | 'refresh_token_guest'
    | 'refresh_token_registered'
    | 'site_id'
```

bfeister replied:
@wjhsf This is nice 👌, this either close to (if not exactly) what I was mentioning in the typescript chat we had with @kevinxh last week

bendvc replied:
Looks promising to my TS n00b eyes. Maybe you can contribute this is a separate PR? @wjhsf

bendvc committed:
Rename ServerStorage to MemoryStorage, allow for use on browser.

- Also use new assignment syntax for private variable.

bendvc committed:
Update packages/commerce-sdk-react/src/auth/storage.ts

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Merge branch 'namespace-auth-storage' of https://github.com/SalesforceCommerceCloud/pwa-kit into namespace-auth-storage

bendvc committed:
Fix some lint errors

bendvc committed:
Rename StorageTypes type to StorageType

bendvc committed:
Add 'memory' to the StorageType type

bendvc committed:
Update packages/commerce-sdk-react/src/auth/storage.test.ts

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Clean up tests...

- Make `onClient` a util so I can mock it.
- Make MemoryStorage use global map by default to maintain server behaviour
- Add test for validating above behaviour

bendvc committed:
Fix linting.

bendvc committed:
Merge branch 'develop' into namespace-auth-storage

bendvc committed:
Update storage.test.ts

kevinxh commented on packages/commerce-sdk-react/src/auth/index.test.ts:
```diff
@@ -246,4 +223,31 @@ describe('Auth', () => {
+        // @ts-expect-error private property
+        expect([...authB.stores['memory'].map.keys()]).toEqual([`siteA_cc-nx-g`, `siteB_cc-nx-g`])
+
+        // Set mock value back to expected.
+        // @ts-expect-error read-only property
+        utils.onClient = () => true
```
I believe jest has util functions to restore mocks, I think it's `jest.restoreAllMocks()` (or maybe `jest.resetAllMocks()`).

We could do that in afterEach to DRY it.
```
afterEach(() => {
  jest.restoreAllMocks()
})
```

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -41,33 +68,62 @@ export class CookieStorage implements BaseStorage {
+const globalMap = new Map()
+
+export class MemoryStorage extends BaseStorage {
+    // private map = new Map<string, string>()
```
```suggestion
```

wjhsf commented on packages/commerce-sdk-react/src/auth/storage.ts:
```diff
@@ -41,33 +68,62 @@ export class CookieStorage implements BaseStorage {
+    constructor(options?: MemoryStorageOptions) {
+        super(options)
+
+        this.map = options?.sharedContext ? globalMap : new Map()
```
Re: https://github.com/SalesforceCommerceCloud/pwa-kit/pull/911#discussion_r1087152328

Our use case is to use a shared context. Is our use case common enough that it should be the default, or should we default to isolated contexts?

bendvc replied:
I chose the default to be isolated maps, as I think it's the most natural. Is it weird that our only use of this is not the default.. maybe.. but I think that our use case of wanting global storage from lambda run to another is a little wweird.

kevinxh commented:
What do you think if we have a test case that looks something like below, to ensure the multisite feature works...

warning: pseudo code 😛 

```jsx
test('Multi-site support', () => {
  const {rerender} = render(<CommerceApiProvider siteId={'en'} />)
  
  // expect storage keys are pre-fixed with siteId EN, pseudocode
  expect(storage.keys).has('en-access_token')

  // re-render the provider with a different site id
  rerender(<CommerceApiProvider siteId={'fr'} />)

  // old access token still exists and value do not change
  expect(storage.keys).has('en-access_token')

  // new access token for FR
  expect(storage.keys).has('fr-access_token')
})
```

bendvc committed:
Update packages/commerce-sdk-react/src/auth/storage.ts

Co-authored-by: Will Harney <62956339+wjhsf@users.noreply.github.com>

bendvc committed:
Merge branch 'develop' into namespace-auth-storage
